Wildcard support for creation in ReplacementTransformer

KnVerey opened the pull request:
⚠️  Behaviour change: in kustomize <v4.5.7, https://github.com/kubernetes-sigs/kustomize/issues/4761 silently ignored targets with out of range indexes (wrongly, IMO!). Kustomize = v4.5.7 panics as described in that issue (also very bad!). My previous PR turned the panic into an error (fine). This one will turn the error into doing what they asked for: error if create is disabled, or creating the fields one by one if create is enabled. The latter would be an output change for a user coming from v4.5.6 or earlier, but I would argue this is fixing something that's clearly a bug.

---

* Makes the Replacements transformer support the use of wildcards in conjunction with the Create option.
* Also makes the Replacements transformer support appending list elements (both "primitive" and map-based elements).
* Does both by implementing these features on PathMatcher, and switching to it. I tried several other things first, which  you can still see in the earlier commits if interested.
  * In working on this, I discovered an existing bug where the Create option cannot work with multiple matches, which is possible without a wildcard (e.g. if you match on container image, that may not be unique in a valid object). That bug is demonstrated in the regression test in this commit: https://github.com/kubernetes-sigs/kustomize/pull/4886/commits/273fdf04439da8b9a9ff28dd62f8e49250962b32. Currently, it silently uses the first one. The other approaches I tried would at best not be able to fix this, which is what led me to conclude modifying and using PathMatcher was necessary.
* Initially, I also added support for `-` meaning "append" without knowing the index in advance. However, the example in https://github.com/kubernetes-sigs/kustomize/issues/4469 made me think the semantics deserve separate consideration (if you use hyphens in that example, you'll append two new things, which is not the intent). Plus PathGetter uses `-` to mean "last", with untested/unclear behaviour on create. We can still do something like this if there's demand for it; I just decided it shouldn't be part of this PR.

Fixes https://github.com/kubernetes-sigs/kustomize/issues/4561
Fixes https://github.com/kubernetes-sigs/kustomize/issues/4469
Fixes https://github.com/kubernetes-sigs/kustomize/issues/1493

KnVerey commented on kyaml/yaml/match.go:
```diff
@@ -43,10 +47,18 @@ type PathMatcher struct {
-	val         *RNode
-	field       string
-	matchRegex  string
-	indexNumber int
```
This field is internal and was set in one place but unused.

natasha41575 reviewed (commented):
The change to PathMatcher is very well done, thanks!

lgtm, couple of minor comments

KnVerey commented:
Updated! I also realized the "find or create" error wasn't showing up in any tests, so I added another one.

> The change to PathMatcher is very well done, thanks!

Thanks! It took me a while to get there, as the commit history still shows. 😅 

/label tide/merge-method-squash

natasha41575 reviewed (commented):
/lgtm
